bson/src/main/org/bson/codecs/EncoderContext.java

```diff
@@ -25,8 +25,10 @@
  * @since 3.0
  */
 public final class EncoderContext {
-
-    private static final EncoderContext DEFAULT_CONTEXT = EncoderContext.builder().build();
+    /**
+     * The default immutable {@link EncoderContext}. Its method {@link #isEncodingCollectibleDocument()} returns {@code false}.
+     */
+    public static final EncoderContext DEFAULT_CONTEXT = EncoderContext.builder().build();
```

Comment:
Please pay attention to this change. Maybe there are reasons why we may not want to expose this field as part of the public API.

I was unable to understand the usage of `EncoderContext` and would appreciate if you guys could help we with answering the following questions:
1. What is the purpose of `EncoderContext.getChildContext`? It returns a final reference `DEFAULT_CONTEXT` to an immutable object, which means that currently this method is futile.
2. The method `EncoderContext.encodeWithChildContext` appears to be futile for the same reason as specified in 1: it always uses `DEFAULT_CONTEXT`.
3. As far as I can tell, currently the class `EncoderContext` could be removed with replacing invocations of the setter method `EncoderContext.Builder.isEncodingCollectibleDocument` with the values currently passed as its arguments (some method signatures and potentially field types may need to be updated).

Reply:
EncoderContext / DecodeContext were invisaged so codecs could understand more about how to encode/decode the data they have and specialize if needed.

> 1.What is the purpose of EncoderContext.getChildContext? It returns a final reference DEFAULT_CONTEXT to an immutable object, which means that currently this method is futile.

Currently, it signals you are encoding a nested field / document so don't need to add a _id field.

> 2. The method EncoderContext.encodeWithChildContext appears to be futile for the same reason as specified in 1: it always uses DEFAULT_CONTEXT.

Its the method used to signal you are encoding a nested field.

> 3. As far as I can tell, currently the class EncoderContext could be removed with replacing invocations of the setter method EncoderContext.Builder.isEncodingCollectibleDocument with the values currently passed as its arguments (some method signatures and potentially field types may need to be updated).

It would be a major breaking API change to remove it, requiring all Codec implementations to be changed. So all internal Codecs and more importantly all userland Codecs would have to change. That in itself is prohibitive and doesn't provide added value to do so.

I think what you are really want to know is "_Why does EncoderContext exist and why was it designed this way?_"

The context pattern is extendable for future unknown requirements when processing data to / from bson. Extending a context doesn't require any major version changes. One example where a change has happened in the past has been for Pojo discriminator support in the DecoderContext.  

The success of the Codec and BsonWriter/Reader design has meant that contexts haven't required to be extended. However, in the past flexibility and user customizability for the contexts has been considered / requested for easier integration into other libraries but so far nothing has been added for external users.

Reply:
@rozza Thank you so much for the detailed explanation! 👍